### PR TITLE
Mean output flows

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/935.

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
To do:
- incorporate https://github.com/Deltares/Ribasim/pull/1155
- At test of mean flow output

File: core/src/parameter.jl [hunk 1]
```diff
@@ -558,8 +558,13 @@ struct Parameters{T, C1, C2}
             edges_source::Dict{Int, Set{EdgeMetadata}},
             flow_dict::Dict{Tuple{NodeID, NodeID}, Int},
             flow::T,
+            flow_prev::Vector{Float64},
+            flow_integrated::Vector{Float64},
             flow_vertical_dict::Dict{NodeID, Int},
             flow_vertical::T,
+            flow_vertical_prev::Vector{Float64},
+            flow_vertical_integrated::Vector{Float64},
```
Comment (visr) [inline core/src/parameter.jl:566]: I think it's fine to leave vertical flows out of this PR. I think we should remove them  altogether. For instance for a Basin, we'll need to integrate the separate vertical fluxes and put those in the Basin struct. Or do you rely on them in allocation?

Comment (SouthEndMusic) [reply core/src/parameter.jl:566]: I think it's equally fine if allocation obtains the mean vertical fluxes from the basin struct.

Comment (visr) [reply core/src/parameter.jl:566]: Ok yeah, I realize it wouldn't make sense to have both instantaneous and mean flows in flows.arrow, but I figured it would be ok if we follow up quickly. I'm fine either way!

File: core/src/callback.jl [hunk 2]
```diff
@@ -379,10 +405,17 @@ end
 
 "Copy the current flow to the SavedValues"
 function save_flow(u, t, integrator)
-    vcat(
-        get_tmp(integrator.p.graph[].flow_vertical, 0.0),
-        get_tmp(integrator.p.graph[].flow, 0.0),
-    )
+    (; graph) = integrator.p
+    (; flow_integrated, flow_vertical_integrated, tprev_flow_save) = graph[]
+    Δt = t - tprev_flow_save[1]
```
Comment (SouthEndMusic) [inline core/src/callback.jl:410]: I don't like the length 1 vector `tprev_flow_save` very much, but we still need something mutable like this because if all timesteps are saved the interval between saves is not constant. Or is there a way to obtain this time interval in a different way?

File: core/test/time_test.jl [hunk 3]
```diff
@@ -29,5 +29,5 @@
     t = Ribasim.seconds_since.(flow_1_to_2.time, model.config.starttime)
     flow_expected = @. 1 + sin(0.5 * π * (t - t[1]) / (t[end] - t[1]))^2
     # some difference is expected since the modeled flow is for the period up to t
-    @test isapprox(flow_added_1, flow_expected, rtol = 0.005)
+    @test isapprox(flow_added_1, flow_expected, rtol = 0.01)
```
Comment (SouthEndMusic) [inline core/test/time_test.jl:32]: Since the saved flows are no longer instantaneous, they should not be used for comparing against exact solutions any further

File: core/src/parameter.jl [hunk 4]
```diff
@@ -558,8 +558,13 @@ struct Parameters{T, C1, C2}
             edges_source::Dict{Int, Set{EdgeMetadata}},
             flow_dict::Dict{Tuple{NodeID, NodeID}, Int},
             flow::T,
+            flow_prev::Vector{Float64},
+            flow_integrated::Vector{Float64},
```
Comment (SouthEndMusic) [inline core/src/parameter.jl:562]: I don't think the integrated flows have to be differentiated

File: core/test/time_test.jl (hunk 3 above)
Comment (visr) [reply core/test/time_test.jl:32]: Yeah that's a bit unfortunate. I guess as long as `saveat` is short enough and the tolerance large enough we can keep using it?

File: core/src/callback.jl (hunk 2 above)
Comment (visr) [reply core/src/callback.jl:410]: Perhaps if we save all timesteps we use `integrator.tprev`, and otherwise the dt is constant.

File: core/test/time_test.jl (hunk 3 above)
Comment (SouthEndMusic) [reply core/test/time_test.jl]: Turns out we can, I wrote this before I noticed that initially my averaged flows were all wrong (:

**Comment by SouthEndMusic** [comment]
To do:
- [ ] Remove horizontal flows from flow output
- [ ] Update docstrings and documentation

File: core/src/callback.jl [hunk 5]
```diff
@@ -379,10 +425,26 @@ end
 
 "Copy the current flow to the SavedValues"
 function save_flow(u, t, integrator)
-    vcat(
-        get_tmp(integrator.p.graph[].flow_vertical, 0.0),
-        get_tmp(integrator.p.graph[].flow, 0.0),
-    )
+    (; tprev, p) = integrator
```
Comment (visr) [inline core/src/callback.jl:428]: There is actually also `dt` in `integrator`, does that help the edge cases a bit?

File: core/test/run_models_test.jl [hunk 6]
```diff
@@ -476,3 +476,28 @@ end
...
+        model = Ribasim.run(config)
+        df = DataFrame(Ribasim.flow_table(model))
+        return filter(
+            [:from_node_id, :to_node_id] => (from, to) -> from === 3 && to === 2,
+            df,
+        ).flow_rate
+    end
+    @test all(get_flow(true, 86400.0) .≈ 1.0)
+    @test all(get_flow(true, 0.0) .≈ 1.0)
+    @test all(get_flow(true, Inf) .≈ 1.0)
```
Comment (visr) [inline core/test/run_models_test.jl:499]: Would be good to also test the length of `tsaves` here, to make sure they are in fact different. And also test `saveat` values that don't fit n times in the tspan, to trigger the remainder path.

File: core/src/callback.jl [hunk 7]
```diff
@@ -377,12 +422,34 @@ function set_control_params!(p::Parameters, node_id::NodeID, control_state::Stri
...
+            iszero(rem) ? saveat : rem
+        end
+    end
+
+    mean_flow_vertical = flow_vertical_integrated / Δt
+    mean_flow = flow_integrated / Δt
+
+    fill!(flow_vertical_integrated, 0.0)
+    fill!(flow_integrated, 0.0)
+    return vcat(mean_flow_vertical, mean_flow)
```
Comment (visr) [inline core/src/callback.jl:452]: Can you vcat the integrated flows first and then divide inplace? That way we are sure to not have more then one array allocation here.

File: core/src/callback.jl [hunk 8]
```diff
@@ -87,6 +92,46 @@ function create_callbacks(
...
+            flow_vertical
+        else
+            0.5 * (flow_vertical + flow_vertical_prev)
+        end
+
+    @. flow_integrated += flow_effective * dt
+    @. flow_vertical_integrated += flow_vertical_effective * dt
+
+    copyto!(flow_prev, flow)
+    copyto!(flow_vertical_prev, flow_vertical)
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/callback.jl:131]: Nice suggestion, the result also looks cleaner

File: core/src/config.jl [hunk 9]
```diff
@@ -233,19 +233,26 @@ end
...
 function convert_saveat(saveat::Float64, t_end::Float64)::Union{Float64, Vector{Float64}}
+    errors = false
     if iszero(saveat)
         # every step
-        Float64[]
+        saveat = Float64[]
     elseif saveat == Inf
         # only the start and end
-        [0.0, t_end]
+        saveat = [0.0, t_end]
```
Comment (visr) [inline core/src/config.jl:242]: Would it not be simpler to return `t_end` (or perhaps `[t_end]`) here instead of filtering the 0.0 out for the SavingCallback?

File: core/test/run_models_test.jl [hunk 10]
```diff
@@ -476,3 +476,70 @@ end
...
+    @test ispath(toml_path)
+    function get_flow(solver_dt::Union{Float64, Nothing}, solver_saveat::Float64)
+        config = Ribasim.Config(toml_path; solver_dt, solver_saveat)
+        model = Ribasim.run(config)
+        df = DataFrame(Ribasim.flow_table(model))
+        return filter(
+            [:from_node_id, :to_node_id] => (from, to) -> from === 3 && to === 2,
+            df,
+        ).flow_rate,
+        Ribasim.tsaves(model)
```
Comment (visr) [inline core/test/run_models_test.jl:494]:
```suggestion
        flow = filter(
            [:from_node_id, :to_node_id] => (from, to) -> from === 3 && to === 2,
            df,
        ).flow_rate,
        return flow, Ribasim.tsaves(model)
```

**Review by visr — APPROVED** [review]

File: core/src/config.jl (hunk 9 above)
Comment (SouthEndMusic) [reply core/src/config.jl:242]: I don't know whether that's desirable since `saveat` is also used for the save subgrid level callback.

Comment (visr) [reply core/src/config.jl:242]: Yeah I was thinking for that we can use save start = true
